starboard/nplb/posix_compliance/posix_pipe_test.cc

```diff
@@ -172,16 +185,36 @@ TEST(PosixPipeTest, ReadFromEmptyPipeCreatedWithNonBlockFlagDoesNotBlock) {
   close(pipe_fds[1]);
 }
 
-void* DoDelayedRead(void* parameter) {
+void* DoReadOnTimeout(void* parameter) {
   TestContext* context = static_cast<TestContext*>(parameter);
 
-  // TODO: b/412690906 - Consider using sem_timedwait() or
-  // pthread_cond_timedwait() instead so that the test can avoid a forced delay
-  // in the passing case.
-  sleep(context->timeout_s);
-  pthread_mutex_lock(&context->out_of_time_mutex);
+  int64_t timeout_time_us =
+      starboard::CurrentPosixTime() + context->timeout_duration_us;
+  struct timespec timeout_time = {0};
+  timeout_time.tv_sec = timeout_time_us / 1'000'000;
+  timeout_time.tv_nsec = (timeout_time_us % 1'000'000) * 1000;
+
+  bool assume_main_thread_blocked_on_write = false;
+
+  EXPECT_EQ(pthread_mutex_lock(&context->had_expected_failed_write_mutex), 0);
+  while (!context->had_expected_failed_write) {
```

Comment:
Is it possible for this to be an infinite loop if the `write` call doesn't return -1 or EAGAIN? I could see us potentially hanging on the pthread_join call